#!/usr/bin/env python3
"""
ç®€å•çš„ gRPC æœåŠ¡å™¨ - å®ç° SiteHealthService
å¿«é€Ÿå¯åŠ¨ç”¨äºæµ‹è¯•
"""

import grpc
from concurrent import futures
import sys
import os

# æ·»åŠ ç”Ÿæˆçš„ä»£ç è·¯å¾„
sys.path.append(os.path.join(os.path.dirname(__file__), 'python_generated'))

try:
    from site_health_pb2 import *
    from site_health_pb2_grpc import *
    from google.protobuf.empty_pb2 import Empty
    from google.protobuf.timestamp_pb2 import Timestamp
    from types_pb2 import HealthStatus
    print("âœ“ æˆåŠŸå¯¼å…¥ proto æ–‡ä»¶")
except ImportError as e:
    print(f"âœ— å¯¼å…¥å¤±è´¥: {e}")
    print("è¯·å…ˆè¿è¡Œ: ./generate_dart_grpc.sh")
    sys.exit(1)

class SimpleSiteHealthService(SiteHealthServiceServicer):
    """ç®€å•çš„åœºåœ°å¥åº·æœåŠ¡å®ç°"""
    
    def GetHealthStatistics(self, request, context):
        """è·å–å¥åº·ç»Ÿè®¡ä¿¡æ¯"""
        print("ğŸ“Š æ”¶åˆ° GetHealthStatistics è¯·æ±‚")
        
        # åˆ›å»ºæ¨¡æ‹Ÿæ•°æ®
        stats = HealthStatistics(
            total_nodes=150,
            normal_nodes=120,
            warning_nodes=20,
            error_nodes=8,
            critical_nodes=2,
            offline_nodes=0,
            overall_health_score=85.5,
        )
        
        # æ·»åŠ æ¯å°è½¦çš„å¼‚å¸¸ç»Ÿè®¡
        for car_id in range(1, 9):
            exception_count = 0 if car_id <= 5 else (car_id - 5)
            stats.per_car_exception_counts.append(
                PerCarExceptionCount(
                    car_id=car_id,
                    exception_count=exception_count,
                )
            )
        
        print(f"âœ“ è¿”å›æ•°æ®: {stats.total_nodes} ä¸ªèŠ‚ç‚¹")
        return stats
    
    def SubscribeHealthStatus(self, request, context):
        """è®¢é˜…å¥åº·çŠ¶æ€å˜åŒ–ï¼ˆæµå¼ï¼‰"""
        print("ğŸ”„ å®¢æˆ·ç«¯è®¢é˜…å¥åº·çŠ¶æ€")
        import time
        import random
        
        counter = 0
        while context.is_active():
            counter += 1
            # æ¨¡æ‹Ÿå‘é€å¥åº·çŠ¶æ€æ›´æ–°
            info = SiteHealthInfo()
            
            # æ¯5ç§’å‘é€ä¸€æ¬¡æ›´æ–°
            if counter % 2 == 0:
                # å‘é€äºŒç»´ç å¥åº·æ•°æ®
                marker = info.marker_health
                marker.node_id = random.randint(1, 150)
                marker.car_id = random.randint(1, 8)
                marker.health_status = HealthStatus.HEALTH_STATUS_NORMAL if random.random() > 0.3 else HealthStatus.HEALTH_STATUS_WARNING
            
            yield info
            time.sleep(5)
    
    def QueryHealthInfo(self, request, context):
        """æŸ¥è¯¢å¥åº·ä¿¡æ¯"""
        print("ğŸ” æ”¶åˆ° QueryHealthInfo è¯·æ±‚")
        
        response = HealthInfoQueryResponse()
        
        # æ·»åŠ ä¸€äº›æ¨¡æ‹Ÿçš„äºŒç»´ç å¥åº·æ•°æ®
        for i in range(10):
            marker = response.marker_healths.add()
            marker.node_id = 100 + i
            marker.car_id = (i % 8) + 1
            marker.health_status = HealthStatus.HEALTH_STATUS_NORMAL
        
        print(f"âœ“ è¿”å› {len(response.marker_healths)} æ¡æ•°æ®")
        return response


def serve():
    """å¯åŠ¨ gRPC æœåŠ¡å™¨"""
    server = grpc.server(
        futures.ThreadPoolExecutor(max_workers=10),
        options=[
            ('grpc.max_send_message_length', 50 * 1024 * 1024),
            ('grpc.max_receive_message_length', 50 * 1024 * 1024),
        ]
    )
    
    # æ³¨å†ŒæœåŠ¡
    add_SiteHealthServiceServicer_to_server(SimpleSiteHealthService(), server)
    
    # ç›‘å¬æ‰€æœ‰ç½‘ç»œæ¥å£
    port = 50051
    server.add_insecure_port(f'0.0.0.0:{port}')
    
    # å¯åŠ¨æœåŠ¡å™¨
    server.start()
    
    print("=" * 60)
    print("ğŸš€ gRPC æœåŠ¡å™¨å¯åŠ¨æˆåŠŸï¼")
    print("=" * 60)
    print(f"ğŸ“¡ ç›‘å¬ç«¯å£: {port}")
    print(f"ğŸ”— æœ¬åœ°åœ°å€: localhost:{port}")
    print(f"ğŸŒ ç½‘ç»œåœ°å€: 0.0.0.0:{port} (å¯é€šè¿‡IPè®¿é—®)")
    print()
    print("å·²å®ç°çš„æ–¹æ³•:")
    print("  - GetHealthStatistics")
    print("  - SubscribeHealthStatus")
    print("  - QueryHealthInfo")
    print()
    print("æŒ‰ Ctrl+C åœæ­¢æœåŠ¡å™¨")
    print("=" * 60)
    
    try:
        server.wait_for_termination()
    except KeyboardInterrupt:
        print("\n\nğŸ›‘ æ­£åœ¨åœæ­¢æœåŠ¡å™¨...")
        server.stop(0)
        print("âœ“ æœåŠ¡å™¨å·²åœæ­¢")


if __name__ == '__main__':
    serve()
